File: src/frontend/agenda/lib/app/data/model/agenda_compromisso_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:agenda/app/infra/infra_imports.dart';
import 'package:agenda/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:agenda/app/data/domain/domain_imports.dart';

class AgendaCompromissoModel {
	int? id;
	int? idAgendaCategoriaCompromisso;
	int? idColaborador;
	DateTime? dataCompromisso;
	String? hora;
	int? duracao;
	String? tipo;
	String? onde;
	String? descricao;
	List<AgendaNotificacaoModel>? agendaNotificacaoModelList;
	List<AgendaCompromissoConvidadoModel>? agendaCompromissoConvidadoModelList;
	List<ReuniaoSalaEventoModel>? reuniaoSalaEventoModelList;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;
	AgendaCategoriaCompromissoModel? agendaCategoriaCompromissoModel;

	AgendaCompromissoModel({
		this.id,
		this.idAgendaCategoriaCompromisso,
		this.idColaborador,
		this.dataCompromisso,
		this.hora,
		this.duracao,
		this.tipo,
		this.onde,
		this.descricao,
		this.agendaNotificacaoModelList,
		this.agendaCompromissoConvidadoModelList,
		this.reuniaoSalaEventoModelList,
		this.viewPessoaColaboradorModel,
		this.agendaCategoriaCompromissoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'data_compromisso',
		'hora',
		'duracao',
		'tipo',
		'onde',
		'descricao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Data Compromisso',
		'Hora',
		'Duracao',
		'Tipo',
		'Onde',
		'Descricao',
	];

	AgendaCompromissoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idAgendaCategoriaCompromisso = jsonData['idAgendaCategoriaCompromisso'];
		idColaborador = jsonData['idColaborador'];
		dataCompromisso = jsonData['dataCompromisso'] != null ? DateTime.tryParse(jsonData['dataCompromisso']) : null;
		hora = jsonData['hora'];
		duracao = jsonData['duracao'];
		tipo = AgendaCompromissoDomain.getTipo(jsonData['tipo']);
		onde = jsonData['onde'];
		descricao = jsonData['descricao'];
		agendaNotificacaoModelList = (jsonData['agendaNotificacaoModelList'] as Iterable?)?.map((m) => AgendaNotificacaoModel.fromJson(m)).toList() ?? [];
		agendaCompromissoConvidadoModelList = (jsonData['agendaCompromissoConvidadoModelList'] as Iterable?)?.map((m) => AgendaCompromissoConvidadoModel.fromJson(m)).toList() ?? [];
		reuniaoSalaEventoModelList = (jsonData['reuniaoSalaEventoModelList'] as Iterable?)?.map((m) => ReuniaoSalaEventoModel.fromJson(m)).toList() ?? [];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
		agendaCategoriaCompromissoModel = jsonData['agendaCategoriaCompromissoModel'] == null ? AgendaCategoriaCompromissoModel() : AgendaCategoriaCompromissoModel.fromJson(jsonData['agendaCategoriaCompromissoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idAgendaCategoriaCompromisso'] = idAgendaCategoriaCompromisso != 0 ? idAgendaCategoriaCompromisso : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['dataCompromisso'] = dataCompromisso != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataCompromisso!) : null;
		jsonData['hora'] = hora;
		jsonData['duracao'] = duracao;
		jsonData['tipo'] = AgendaCompromissoDomain.setTipo(tipo);
		jsonData['onde'] = onde;
		jsonData['descricao'] = descricao;
		
		var agendaNotificacaoModelLocalList = []; 
		for (AgendaNotificacaoModel object in agendaNotificacaoModelList ?? []) { 
			agendaNotificacaoModelLocalList.add(object.toJson); 
		}
		jsonData['agendaNotificacaoModelList'] = agendaNotificacaoModelLocalList;
		
		var agendaCompromissoConvidadoModelLocalList = []; 
		for (AgendaCompromissoConvidadoModel object in agendaCompromissoConvidadoModelList ?? []) { 
			agendaCompromissoConvidadoModelLocalList.add(object.toJson); 
		}
		jsonData['agendaCompromissoConvidadoModelList'] = agendaCompromissoConvidadoModelLocalList;
		
		var reuniaoSalaEventoModelLocalList = []; 
		for (ReuniaoSalaEventoModel object in reuniaoSalaEventoModelList ?? []) { 
			reuniaoSalaEventoModelLocalList.add(object.toJson); 
		}
		jsonData['reuniaoSalaEventoModelList'] = reuniaoSalaEventoModelLocalList;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
		jsonData['agendaCategoriaCompromissoModel'] = agendaCategoriaCompromissoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idAgendaCategoriaCompromisso = plutoRow.cells['idAgendaCategoriaCompromisso']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		dataCompromisso = Util.stringToDate(plutoRow.cells['dataCompromisso']?.value);
		hora = plutoRow.cells['hora']?.value;
		duracao = plutoRow.cells['duracao']?.value;
		tipo = plutoRow.cells['tipo']?.value != '' ? plutoRow.cells['tipo']?.value : 'Pessoal';
		onde = plutoRow.cells['onde']?.value;
		descricao = plutoRow.cells['descricao']?.value;
		agendaNotificacaoModelList = [];
		agendaCompromissoConvidadoModelList = [];
		reuniaoSalaEventoModelList = [];
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
		agendaCategoriaCompromissoModel = AgendaCategoriaCompromissoModel();
		agendaCategoriaCompromissoModel?.nome = plutoRow.cells['agendaCategoriaCompromissoModel']?.value;
	}	

	AgendaCompromissoModel clone() {
		return AgendaCompromissoModel(
			id: id,
			idAgendaCategoriaCompromisso: idAgendaCategoriaCompromisso,
			idColaborador: idColaborador,
			dataCompromisso: dataCompromisso,
			hora: hora,
			duracao: duracao,
			tipo: tipo,
			onde: onde,
			descricao: descricao,
			agendaNotificacaoModelList: agendaNotificacaoModelListClone(agendaNotificacaoModelList!),
			agendaCompromissoConvidadoModelList: agendaCompromissoConvidadoModelListClone(agendaCompromissoConvidadoModelList!),
			reuniaoSalaEventoModelList: reuniaoSalaEventoModelListClone(reuniaoSalaEventoModelList!),
		);			
	}

	agendaNotificacaoModelListClone(List<AgendaNotificacaoModel> agendaNotificacaoModelList) { 
		List<AgendaNotificacaoModel> resultList = [];
		for (var agendaNotificacaoModel in agendaNotificacaoModelList) {
			resultList.add(
				AgendaNotificacaoModel(
					id: agendaNotificacaoModel.id,
					idAgendaCompromisso: agendaNotificacaoModel.idAgendaCompromisso,
					dataNotificacao: agendaNotificacaoModel.dataNotificacao,
					hora: agendaNotificacaoModel.hora,
					tipo: agendaNotificacaoModel.tipo,
				)
			);
		}
		return resultList;
	}

	agendaCompromissoConvidadoModelListClone(List<AgendaCompromissoConvidadoModel> agendaCompromissoConvidadoModelList) { 
		List<AgendaCompromissoConvidadoModel> resultList = [];
		for (var agendaCompromissoConvidadoModel in agendaCompromissoConvidadoModelList) {
			resultList.add(
				AgendaCompromissoConvidadoModel(
					id: agendaCompromissoConvidadoModel.id,
					idAgendaCompromisso: agendaCompromissoConvidadoModel.idAgendaCompromisso,
					idColaborador: agendaCompromissoConvidadoModel.idColaborador,
				)
			);
		}
		return resultList;
	}

	reuniaoSalaEventoModelListClone(List<ReuniaoSalaEventoModel> reuniaoSalaEventoModelList) { 
		List<ReuniaoSalaEventoModel> resultList = [];
		for (var reuniaoSalaEventoModel in reuniaoSalaEventoModelList) {
			resultList.add(
				ReuniaoSalaEventoModel(
					id: reuniaoSalaEventoModel.id,
					idAgendaCompromisso: reuniaoSalaEventoModel.idAgendaCompromisso,
					idReuniaoSala: reuniaoSalaEventoModel.idReuniaoSala,
					dataReserva: reuniaoSalaEventoModel.dataReserva,
				)
			);
		}
		return resultList;
	}

	
}

File: src/backend/agenda/src/model/agenda_compromisso_model.py
from src import db
from src.model.agenda_notificacao_model import AgendaNotificacaoModel
from src.model.agenda_compromisso_convidado_model import AgendaCompromissoConvidadoModel
from src.model.reuniao_sala_evento_model import ReuniaoSalaEventoModel
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel
from src.model.agenda_categoria_compromisso_model import AgendaCategoriaCompromissoModel


class AgendaCompromissoModel(db.Model):
    __tablename__ = 'agenda_compromisso'

    id = db.Column(db.Integer, primary_key=True)
    data_compromisso = db.Column(db.DateTime)
    hora = db.Column(db.String(8))
    duracao = db.Column(db.Integer)
    tipo = db.Column(db.String(1))
    onde = db.Column(db.String(100))
    descricao = db.Column(db.String(100))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))
    id_agenda_categoria_compromisso = db.Column(db.Integer, db.ForeignKey('agenda_categoria_compromisso.id'))

    agenda_notificacao_model_list = db.relationship('AgendaNotificacaoModel', lazy='dynamic')
    agenda_compromisso_convidado_model_list = db.relationship('AgendaCompromissoConvidadoModel', lazy='dynamic')
    reuniao_sala_evento_model_list = db.relationship('ReuniaoSalaEventoModel', lazy='dynamic')
    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])
    agenda_categoria_compromisso_model = db.relationship('AgendaCategoriaCompromissoModel', foreign_keys=[id_agenda_categoria_compromisso])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_agenda_categoria_compromisso = data.get('idAgendaCategoriaCompromisso')
        self.id_colaborador = data.get('idColaborador')
        self.data_compromisso = data.get('dataCompromisso')
        self.hora = data.get('hora')
        self.duracao = data.get('duracao')
        self.tipo = data.get('tipo')
        self.onde = data.get('onde')
        self.descricao = data.get('descricao')

    def serialize(self):
        return {
            'id': self.id,
            'idAgendaCategoriaCompromisso': self.id_agenda_categoria_compromisso,
            'idColaborador': self.id_colaborador,
            'dataCompromisso': self.data_compromisso.isoformat(),
            'hora': self.hora,
            'duracao': self.duracao,
            'tipo': self.tipo,
            'onde': self.onde,
            'descricao': self.descricao,
            'agendaNotificacaoModelList': [agenda_notificacao_model.serialize() for agenda_notificacao_model in self.agenda_notificacao_model_list],
            'agendaCompromissoConvidadoModelList': [agenda_compromisso_convidado_model.serialize() for agenda_compromisso_convidado_model in self.agenda_compromisso_convidado_model_list],
            'reuniaoSalaEventoModelList': [reuniao_sala_evento_model.serialize() for reuniao_sala_evento_model in self.reuniao_sala_evento_model_list],
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
            'agendaCategoriaCompromissoModel': self.agenda_categoria_compromisso_model.serialize() if self.agenda_categoria_compromisso_model else None,
        }